adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -197,6 +196,37 @@ protected void doStart(final Promise<Void> startPromise) {
         .onComplete(startPromise);
     }
 
+    /**
+     * Handles any operations that should be invoked as part of the authentication process after the credentials got
+     * determined and before they get validated. Can be used to perform checks using the credentials and tenant
+     * information before the potentially expensive credentials validation is done
+     * <p>
+     * The default implementation updates the trace sampling priority in the execution context tracing span.
+     * <p>
+     * Subclasses should override this method in order to perform additional operations after calling this super method.
+     *
+     * @param credentials The credentials.
+     * @param executionContext The execution context, including the TenantObject.
+     * @return A future indicating the outcome of the operation. A failed future will fail the authentication attempt.
+     */
+    protected Future<Void> handleBeforeCredentialsValidation(final DeviceCredentials credentials,
+            final SaslResponseContext executionContext) {
+
+        final String tenantId = credentials.getTenantId();
+        final Span span = executionContext.getTracingSpan();
+        final String authId = credentials.getAuthId();
+
+        return getTenantConfiguration(tenantId, span.context())
+                .compose(tenantObject -> {
+                    TracingHelper.setDeviceTags(span, tenantId, null, authId);
+                    final OptionalInt traceSamplingPriority = TenantTraceSamplingHelper.applyTraceSamplingPriority(
+                            tenantObject, authId, span);
+                    executionContext.getProtonConnection().attachments().set(
```

Comment:
FMPOV we should not rely on the internal structure of the SaslResponseContext here but instead simply invoke something like *SaslResponseContext.setTraceSamplingPriority(TenantObject)* to improve encapsulation ...

Reply:
Yes, however it's not so straightforward at the moment because the places where you would want to call `getTraceSamplingPriority` don't have the `SaslResponseContext` object available.
I wanted to replace the various different objects being put into the connection attachments with one single "ConnectionContext" object anyway. I would rather do this in a follow-up PR then though.